### PR TITLE
ARROW-12120: [Rust] Generate random arrays and batches

### PR DESCRIPTION
**Pull request by nevi-me**
This generates random batches and arrays, mostly to assist with tests and benchmarks.

### CODE REVIEW

File: rust/arrow/benches/equal.rs [hunk 1]
```diff
@@ -38,10 +38,10 @@ fn add_benchmark(c: &mut Criterion) {
     let arr_a_nulls = create_primitive_array::<Float32Type>(512, 0.5);
     c.bench_function("equal_nulls_512", |b| b.iter(|| bench_equal(&arr_a_nulls)));
 
-    let arr_a = create_string_array(512, 0.0);
+    let arr_a = create_string_array::<i32>(512, 0.0);
```
Comment (Dandandan) [inline rust/arrow/benches/equal.rs:41]: Not part of the PR - but I notice this benchmark is strange / not very representative as it is testing two of the same arrays to be equal.

File: rust/arrow/src/util/data_gen.rs [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+        Int16 => Arc::new(create_primitive_array::<Int16Type>(size, null_density)),
+        Int32 => Arc::new(create_primitive_array::<Int32Type>(size, null_density)),
+        Int64 => Arc::new(create_primitive_array::<Int64Type>(size, null_density)),
+        UInt8 => Arc::new(create_primitive_array::<UInt8Type>(size, null_density)),
+        UInt16 => Arc::new(create_primitive_array::<UInt16Type>(size, null_density)),
+        UInt32 => Arc::new(create_primitive_array::<UInt32Type>(size, null_density)),
+        UInt64 => Arc::new(create_primitive_array::<UInt64Type>(size, null_density)),
+        Float16 => {
+            return Err(ArrowError::NotYetImplemented(
+                "Float16 is not implememted".to_string(),
```
Comment (Dandandan) [inline rust/arrow/src/util/data_gen.rs:83]:
```suggestion
                "Float16 is not implemented".to_string(),
```

Comment (nevi-me) [reply rust/arrow/src/util/data_gen.rs:83]: Sorry, that's how you say it in an African accent 😄

File: rust/arrow/benches/equal.rs (hunk 1 above)
Comment (nevi-me) [reply rust/arrow/benches/equal.rs:41]: Yeah, I see what you mean, we can create another array, and compare the 2

Comment (Dandandan) [reply rust/arrow/benches/equal.rs:41]: Yeah, and also the seeds need to be different in that case, earlier I found a benchmark which created two same arrays, but as the code uses the same `42` seed, it still had to equivalent arrays :shrug:

**Review by jorgecarleitao — APPROVED** [review]
LGTM. Good idea. It may also be useful to run microbenches in DataFusion.
